### PR TITLE
ipc_service: Automatically include source files for backends

### PR DESCRIPTION
**Pull request by eobalski**
The purpose of this change is to allow to enable more than one
backend at once.

The outcoming changes include:
- removal of choice option for IPC backends Kconfig
- removal of ZTEST backend
- removal necessity of explicitly specifying backend
- move IPC_SERVICE_NUM_ENDPOINTS_PER_INSTANCE option
  to Kconfig.rpmsg (+rename)
- automatical source file inclusion for chosen backend (now depends
  on DT configuration)
- clean up of samples and Kconfig files that were explicitly chosing
  STATIC_VRINGS as backend

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

### CODE REVIEW

**Comment by emob-nordic** [comment]
Hi @kapi-no. I would like to kindly ask you to have a look here. I'm changing some things you've recently added. I think we don't need to explicitly select backends and this could be done as presented.
@e-rk if possible, could you verify if this suits you as well and if it affect your work in any significant way?
@carlocaione Those are the changes we discussed offline some time ago, please fell free to speak out.

**Comment by emob-nordic** [comment]
> can you please split the BT changes from the general IPC ones in two separate commits?

IPC service won't work on 1st commit then. Are you sure about that? This change is pretty simple. I can do it but is there much need for that?

File: samples/subsys/ipc/ipc_service/static_vrings/remote/prj.conf [hunk 1]
```diff
@@ -2,7 +2,6 @@ CONFIG_PRINTK=y
 CONFIG_HEAP_MEM_POOL_SIZE=4096
 
 CONFIG_IPC_SERVICE=y
-CONFIG_IPC_SERVICE_BACKEND_RPMSG=y
```
Comment (nordic-krch) [inline samples/subsys/ipc/ipc_service/static_vrings/remote/prj.conf:5]: this could be split into another commit, since you are only changing the default in kconfig so keeping it here won't break any build, right?

File: subsys/ipc/ipc_service/backends/Kconfig [hunk 2]
```diff
@@ -1,46 +1,37 @@
...
-
-config IPC_SERVICE_REG_BACKEND_PRIORITY
-	int "Initialization priority of modules registering IPC backend"
-	default 46
-	help
-	  The backend must be registered before the endpoint register.
-
 config IPC_SERVICE_RPMSG
 	bool "RPMsg support library"
+	depends on IPC_SERVICE_BACKEND_RPMSG
```
Comment (carlocaione) [inline subsys/ipc/ipc_service/backends/Kconfig:22]: This doesn't actually depends on that backend. The backend is selecting this library, but the library can be selected by multiple backends. So, this is not needed.

File: subsys/ipc/ipc_service/backends/Kconfig [hunk 3]
```diff
@@ -1,46 +1,37 @@
...
+config IPC_SERVICE_REG_BACKEND_PRIORITY
+	int "Initialization priority of modules registering IPC backend"
+	default 46
+	help
+	  The backend must be registered before the endpoint register.
 
 config IPC_SERVICE_BACKEND_RPMSG
 	bool "OpenAMP RPMSG backend with static VRINGs"
 	depends on MBOX
+	default $(dt_compat_enabled,$(DT_COMPAT_ZEPHYR_IPC_OPENAMP_STATIC_VRINGS))
```
Comment (carlocaione) [inline subsys/ipc/ipc_service/backends/Kconfig:15]: This confuses me. This change is implying that to select this backend we must change the DT. Without this change we are implying that we have to modify the defconfig instead. Now, do we have a strong preference for a way or the other?

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: subsys/ipc/ipc_service/backends/Kconfig (hunk 3 above)
Comment (emob-nordic) [reply subsys/ipc/ipc_service/backends/Kconfig:15]: I don't see any need to enforce the users to explicitly enable this option in configuration file. Afaik ipc-service was ment to be DTS configurable and this change actually does that. I just don't see any advantage of not making it this way. If the user adds `zephyr,ipc-openamp-static-vrings` compatible i believe he/she knows what he/she is doing.
This is more of a proposal rather than hard statement. I know you have objections about that, but i don't understand the justification of why not to do like so.

The only argument against such changes i heard in the past was: Kconfig is for software configuration and DT is for hardware configuration. If so - then configuring backend should never be DT configurable in the first place (imo). But we do accept that, so why not go one step further?

File: subsys/ipc/ipc_service/backends/Kconfig [hunk 4]
```diff
@@ -1,46 +1,37 @@
...
 
 config IPC_SERVICE_STATIC_VRINGS
 	bool "Static VRINGs support library"
+	depends on IPC_SERVICE_BACKEND_RPMSG
 	help
 	  "Static VRINGs library"
 
 config IPC_SERVICE_STATIC_VRINGS_ALIGNMENT
 	int "VRINGs alignment"
+	depends on IPC_SERVICE_BACKEND_RPMSG
```
[the remark this answers is not in the document]
Comment (emob-nordic) [reply subsys/ipc/ipc_service/backends/Kconfig:34]: right.

File: subsys/ipc/ipc_service/backends/Kconfig (hunk 3 above)
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/Kconfig:15]: > configuring backend should never be DT configurable in the first place (imo)

Right, but also Linux is going to have OpenAMP nodes in DT, so we are not the only ones :)

That said I don't have a strong opinion on this, so this is fine to me.

**Comment by emob-nordic** [comment]
Force pushed with changes + rebase due to conflicts.
- Removed depends on mentioned by @carlocaione and @nordic-krch  - see reviews above.
- Split removal of configuration options from samples to separate commit as it doesn't break anything.

File: subsys/ipc/ipc_service/backends/Kconfig [hunk 5]
```diff
@@ -1,42 +1,32 @@
...
 # SPDX-License-Identifier: Apache-2.0
 
-choice IPC_SERVICE_BACKEND
-	prompt "IPC service backend"
+DT_COMPAT_ZEPHYR_IPC_OPENAMP_STATIC_VRINGS := zephyr,ipc-openamp-static-vrings
+DT_COMPAT_ZEPHYR_IPC_ICMSG := zephyr,ipc-icmsg
+
+config IPC_SERVICE_REG_BACKEND_PRIORITY
+	int "Initialization priority of modules registering IPC backend"
+	default 46
```
Comment (nordic-krch) [inline subsys/ipc/ipc_service/backends/Kconfig:9]: any chance to have static assert which would verify priority correctness? Something like: https://github.com/zephyrproject-rtos/zephyr/blob/93b0db781eafe9fbff24df6b4d31324c76c6941b/soc/arm/nordic_nrf/nrf53/sync_rtc.c#L310

Comment (emob-nordic) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: Actually i just moved it, i don't know what was the intended relation. @carlocaione what should we put there?

Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: The only relation I can think of is with the MBOX driver that should be probed before the backend. Adding a `BUILD_ASSERT` for that is ok to me.

Comment (emob-nordic) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: uhmmm, why is it `default 46` then? Default MBOX_INIT_PRIORITY is `Kernel:50`. For backend it would be `Kernel:46` now. Lower initialization priority means earlier initialization. Shouldn't it be > 50 then?
@carlocaione

Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: yeah, but that's not really import for the static vrings backend. That backend is initializing the mbox when opening the instance not on init. This change could be useful for other backends that are initializing the mbox during init.

Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: so, yeah, change the priority with respect to the mbox and add the assert even though that makes no difference for the current backend.

**Review by e-rk — DISMISSED** [review]
No regressions observed with ieee802154 echo_client/server samples.

File: subsys/ipc/ipc_service/backends/Kconfig [hunk 6]
```diff
@@ -1,42 +1,32 @@
...
+	help
+	  The backend must be registered before the endpoint register.
 
 config IPC_SERVICE_BACKEND_RPMSG
 	bool "OpenAMP RPMSG backend with static VRINGs"
```
[the remark this answers is not in the document]
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/Kconfig:14]: yes please. Some platforms would want to enable this using defconfig instead.

File: subsys/ipc/ipc_service/backends/Kconfig (hunk 5 above)
Comment (emob-nordic) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: Shall we keep this option at all? I think there is no point in keeping that if any of the backends actually depends on `MBOX_INIT_PRIORITY`. Shall we use `CONFIG_KERNEL_INIT_PRIORITY_DEVICE` for both backends we have now?

Comment (emob-nordic) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: I think there is no good reason to multiply options here if not needed. @nordic-krch  what do you think?

Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/Kconfig:9]: please keep it. The point is that you can have some backend that is depending on some other driver that must be probed before the backend. So you can enforce a probe ordering only by acting on this symbol.

**Comment by emob-nordic** [comment]
Rebased due to conflicts.

**Review by rakons — DISMISSED** [review]

File: subsys/ipc/ipc_service/ipc_service.c [hunk 7]
```diff
@@ -12,6 +12,11 @@
 #include <zephyr.h>
 #include <device.h>
 
+#if defined(CONFIG_MBOX_INIT_PRIORITY)
+BUILD_ASSERT(CONFIG_IPC_SERVICE_REG_BACKEND_PRIORITY > CONFIG_MBOX_INIT_PRIORITY,
+		"IPC service backend must be initialized after MBOX driver.");
+#endif
+
```
Comment (carlocaione) [inline subsys/ipc/ipc_service/ipc_service.c:19]: This should not be here because not necessarily the backend is using the MBOX. This is true for `static_vrings` but it's not a general solution.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: subsys/ipc/ipc_service/ipc_service.c (hunk 7 above)
Comment (emob-nordic) [reply subsys/ipc/ipc_service/ipc_service.c:19]: Ah, yes. Sorry i forgot to move it.

Comment (emob-nordic) [reply subsys/ipc/ipc_service/ipc_service.c:19]: Discussed offline. This assertion is not required for any backend right now. Im removing the last commit.

**Review by nordic-krch — APPROVED** [review]

**Review by e-rk — APPROVED** [review]

**Review by pdunaj — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]
